sgkit/io/dataset.py

```diff
@@ -48,7 +52,34 @@ def save_dataset(
             filters.remove(var_len_str_codec)
             ds[v].encoding["filters"] = filters
 
-    ds.to_zarr(store, **kwargs)
+    if auto_rechunk:
+        # This logic for checking if rechunking is necessary is
+        # taken from xarray/backends/zarr.py#L109.
+        # We can't try to save and catch the error as by that
+        # point the zarr store is non-empty.
+        if any(len(set(chunks[:-1])) > 1 for chunks in ds.chunks.values()) or any(
+            (chunks[0] < chunks[-1]) for chunks in ds.chunks.values()
+        ):
+            # Here we use the max chunk size as the target chunk size as for the commonest
+            # case of subsetting an existing dataset, this will be closest to the original
+            # intended chunk size.
+            ds = ds.chunk(
+                chunks={dim: max(chunks) for dim, chunks in ds.chunks.items()}
+            )
+
+    # Catch unequal chunking errors to provide a more helpful error message
+    try:
+        ds.to_zarr(store, **kwargs)
+    except ValueError as e:
+        if "Zarr requires uniform chunk sizes" in str(
```

Comment:
This feels a bit brittle - what happens when upstream changes the error messages slightly?

Reply:
Yes, I agree, but I think its the best of two bad options. The other option is to copy across the detection code from xarray - but then we can't tell if they have changed the logic. At least this way our tests should fail if xarray changes the message. If they do so, and we haven't updated the end user will see the xarray message as before, which isn't awful, just inconvenient.